### PR TITLE
FIX: properly handle dates when intmult is true

### PR DESCRIPTION
**Pull request by jklymak**
## PR Summary

Currently, if interval_multiples=True (now the default), and a long enough numbers of days is being plotted, the locator finds the 1st and 22nd of the month as dates.  That makes sense if interval_multiples=False, but is silly for interval_multiples=True.  

```python 
dates = np.arange('2001-01-10', '2001-05-23', dtype='datetime64[D]')
y = np.sin(dates.astype(float) / 10)
fig, ax = plt.subplots(constrained_layout=True)

#plt.rcParams['date.converter'] = 'concise'
#plt.rcParams['date.interval_multiples'] = True
ax.plot(dates, y)

plt.show()
```

picks dates 1, and 22 of the month.  This PR fixes that to be 1 and 15, and to not raise a warning:

### Before

Before11

### After

After1

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by jklymak** [comment]
Codecov is made because we never call the warning in the tests.  I'm not sure what case this would actually get called, but could try if folks think its necessary.

**Comment by tacaswell** [comment]
How pathological of an example do you need to trigger the warning?  We have had cases of un-exercised warnings bit-rotting and turning into exceptions due to missing variables / bad formatting / etc so a test would be better, but not if it is going to be a huge amount of work.

**Comment by jklymak** [comment]
OK, this warning comes up if we don't satisfy the maxticks constraint, but then it tells the user to change something they can't change anyhow.  I think we should just drop the warning.

**Comment by QuLogic** [comment]
> but then it tells the user to change something they can't change anyhow.

What do you mean? It's an instance attribute. It's even documented in the class docstring.

**Comment by jklymak** [comment]
Ooops!  Fair enough....

**Comment by jklymak** [comment]
So what happens here is that there is special logic for the `DAILY` frequency for which the highest intervald is 14.  If you have 145 days in your time interval, you would fail the test `num <= interval * (self.maxticks[freq] - 1)` because you'd have too many ticks.  Indeed, you'd expect to get one at 1 Jan, 15 Jan, and 29 Jan.  *However* if interval_multiples is True, we explicitly drop this last tick, so there is only a tick at 1 Jan, 15 Jan, 1 Feb, etc.  So there are *not* actually too many ticks, and reaching this point is fine.  

The compromise I reached is to keep the warning in, but exclude checking it for the case of `interval_multiples=True` and the `DAILY` frequency.

**Comment by jklymak** [comment]
Added extra test (though there was no test for this warning before)

**Comment by jklymak** [comment]
I just assumed that was a flaky test.  Its funny that the test passes py3.6 on Azure, but fails on Travis.  I really don't understand what might be different.  py test versions?

```
with pytest.warns(UserWarning) as rec:
           locs = locator()
>           assert len(rec) == 1
E           assert 5 == 1
E            +  where 5 = len(WarningsChecker(record=True))
```

**Comment by jklymak** [comment]
This test only fails on python 3.6 on Travis with 

```
       with pytest.warns(UserWarning) as rec:
2523            locs = locator()
            print(rec)
>           assert len(rec) == 1
E           assert 5 == 1
E            +  where 5 = len(WarningsChecker(record=True))
lib/matplotlib/tests/test_dates.py:962: AssertionError
----------------------------- Captured stdout call -----------------------------
WarningsChecker(record=True)

```

I'm completely befuddled by this, as the same machinery is used elsewhere in this file with no problem.

**Comment by QuLogic** [comment]
Print all the values in `rec`; I suspect this might be a warning about setting locale for datetimes?
https://github.com/matplotlib/matplotlib/blob/45d27c9fade4c6f184a6b9c94db905b250840f16/lib/matplotlib/testing/__init__.py#L34-L42

**Comment by jklymak** [comment]
Trying that now.  Not sure how you knew that might be the problem, but... 

I'm still not clear why that would only affect this one test set up...

**Comment by QuLogic** [comment]
It's the only warning I know of that sometimes happens in containers; may or may not be the actual problem though.

**Comment by QuLogic** [comment]
Not quite as I guessed:
```
{message : ResourceWarning("unclosed file <_io.BufferedReader name='/usr/share/zoneinfo/UTC'>",), category : 'ResourceWarning', filename : '/home/travis/virtualenv/python3.6.7/lib/python3.6/site-packages/dateutil/tz.py', lineno : 930, line : None}
{message : ResourceWarning("unclosed file <_io.BufferedReader name='/usr/share/zoneinfo/UTC'>",), category : 'ResourceWarning', filename : '/home/travis/virtualenv/python3.6.7/lib/python3.6/site-packages/dateutil/tz.py', lineno : 930, line : None}
{message : ResourceWarning("unclosed file <_io.BufferedReader name='/usr/share/zoneinfo/UTC'>",), category : 'ResourceWarning', filename : '/home/travis/virtualenv/python3.6.7/lib/python3.6/site-packages/dateutil/tz.py', lineno : 930, line : None}
{message : ResourceWarning("unclosed file <_io.BufferedReader name='/usr/share/zoneinfo/UTC'>",), category : 'ResourceWarning', filename : '/home/travis/virtualenv/python3.6.7/lib/python3.6/site-packages/dateutil/tz.py', lineno : 930, line : None}
```

I suggest using `pytest.warns(UserWarning, match="AutoDateLocator was unable")`, and I _think_ it ignores unmatched warnings.

File: doc/api/api_changes_3.4/behaviour.rst [hunk 1]
```diff
@@ -1,2 +1,12 @@
 Behaviour changes
 -----------------
+
+Date locator for DAILY interval now returns middle of month
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The `.matplotlib.dates/AutoDateLocator` has a default of
+`interval_multiples=True` that attempts to align ticks with the start of
```
Comment (QuLogic) [inline doc/api/api_changes_3.4/behaviour.rst:8]:
```suggestion
The `matplotlib.dates.AutoDateLocator` has a default of
``interval_multiples=True`` that attempts to align ticks with the start of
```

**Comment by tacaswell** [comment]
Took the liberty of re-basing.

**Comment by tacaswell** [comment]
Anyone can merge on green.

**Pull request merged**